File: generated_code/cpp/solution_838_1.cpp
#include <iostream>

// Definition for the linked list node.
struct Node {
    int val;
    Node* prev;
    Node* next;
    Node(int v) : val(v), prev(nullptr), next(nullptr) {}
};

/**
 * Doubly Linked List Implementation with Dummy Head and Tail.
 * This approach prevents edge cases related to updating head/tail pointers,
 * making insertion and deletion operations O(1) once the node position is reached.
 * Time Complexity:
 * - get, addAtIndex, deleteAtIndex: O(N) in the worst case (where N is the list size).
 * - addAtHead, addAtTail: O(1) since we maintain direct links via dummy nodes.
 * Space Complexity: O(N) to store the elements.
 */
class MyLinkedList {
private:
    Node* head;
    Node* tail;
    int size;

public:
    // Initializes the MyLinkedList object with dummy head and tail nodes.
    MyLinkedList() {
        head = new Node(-1);
        tail = new Node(-1);
        head->next = tail;
        tail->prev = head;
        size = 0;
    }
    
    // Destructor to prevent memory leaks by freeing allocated nodes.
    ~MyLinkedList() {
        Node* curr = head;
        while (curr != nullptr) {
            Node* nextNode = curr->next;
            delete curr;
            curr = nextNode;
        }
    }
    
    // Get the value of the index-th node. Returns -1 if the index is invalid.
    int get(int index) {
        if (index < 0 || index >= size) {
            return -1;
        }
        
        Node* curr;
        // Optimize search direction based on whether index is closer to head or tail
        if (index < size / 2) {
            curr = head->next;
            for (int i = 0; i < index; ++i) {
                curr = curr->next;
            }
        } else {
            curr = tail->prev;
            for (int i = 0; i < size - 1 - index; ++i) {
                curr = curr->prev;
            }
        }
        return curr->val;
    }
    
    // Add a node of value val before the first element of the linked list.
    void addAtHead(int val) {
        Node* newNode = new Node(val);
        Node* nextNode = head->next;
        
        newNode->next = nextNode;
        newNode->prev = head;
        head->next = newNode;
        nextNode->prev = newNode;
        
        size++;
    }
    
    // Append a node of value val as the last element of the linked list.
    void addAtTail(int val) {
        Node* newNode = new Node(val);
        Node* prevNode = tail->prev;
        
        newNode->next = tail;
        newNode->prev = prevNode;
        prevNode->next = newNode;
        tail->prev = newNode;
        
        size++;
    }
    
    // Add a node of value val before the index-th node. 
    // If index equals size, append to the end. If index > size, do nothing.
    void addAtIndex(int index, int val) {
        if (index < 0 || index > size) {
            return;
        }
        
        if (index == size) {
            addAtTail(val);
            return;
        }
        if (index == 0) {
            addAtHead(val);
            return;
        }
        
        Node* curr;
        // Search from head or tail to find the node currently at 'index'
        if (index < size / 2) {
            curr = head->next;
            for (int i = 0; i < index; ++i) {
                curr = curr->next;
            }
        } else {
            curr = tail->prev;
            for (int i = 0; i < size - 1 - index; ++i) {
                curr = curr->prev;
            }
        }
        
        // Insert newNode before curr
        Node* newNode = new Node(val);
        Node* prevNode = curr->prev;
        
        newNode->next = curr;
        newNode->prev = prevNode;
        prevNode->next = newNode;
        curr->prev = newNode;
        
        size++;
    }
    
    // Delete the index-th node in the linked list, if the index is valid.
    void deleteAtIndex(int index) {
        if (index < 0 || index >= size) {
            return;
        }
        
        Node* curr;
        // Search from head or tail to find the node to delete
        if (index < size / 2) {
            curr = head->next;
            for (int i = 0; i < index; ++i) {
                curr = curr->next;
            }
        } else {
            curr = tail->prev;
            for (int i = 0; i < size - 1 - index; ++i) {
                curr = curr->prev;
            }
        }
        
        // Unlink the node and free memory
        Node* prevNode = curr->prev;
        Node* nextNode = curr->next;
        
        prevNode->next = nextNode;
        nextNode->prev = prevNode;
        
        delete curr;
        size--;
    }
};

File: generated_code/python/solution_838_1.py
# Necessary imports for type hinting and data structures
from typing import Optional

class ListNode:
    """Definition for a doubly-linked list node."""
    def __init__(self, val: int = 0):
        self.val = val
        self.next: Optional[ListNode] = None
        self.prev: Optional[ListNode] = None

class MyLinkedList:
    """
    Implementation of a doubly-linked list with sentinel head and tail nodes.
    Using a doubly-linked list with pseudo-head and pseudo-tail simplifies boundary conditions,
    allowing O(1) insertions at head/tail and efficient O(min(index, size - index)) lookups and modifications.
    """
    def __init__(self):
        self.size = 0
        # Sentinel (dummy) nodes to simplify edge cases (empty list operations)
        self.head = ListNode(0)
        self.tail = ListNode(0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, index: int) -> int:
        # Check for out-of-bounds indices
        if index < 0 or index >= self.size:
            return -1
        
        # Optimize search direction based on which end is closer
        if index + 1 < self.size - index:
            curr = self.head
            for _ in range(index + 1):
                curr = curr.next
        else:
            curr = self.tail
            for _ in range(self.size - index):
                curr = curr.prev
                
        return curr.val

    def addAtHead(self, val: int) -> None:
        # Add node immediately after the sentinel head
        pred, succ = self.head, self.head.next
        
        self.size += 1
        to_add = ListNode(val)
        to_add.prev = pred
        to_add.next = succ
        pred.next = to_add
        succ.prev = to_add

    def addAtTail(self, val: int) -> None:
        # Add node immediately before the sentinel tail
        pred, succ = self.tail.prev, self.tail
        
        self.size += 1
        to_add = ListNode(val)
        to_add.prev = pred
        to_add.next = succ
        pred.next = to_add
        succ.prev = to_add

    def addAtIndex(self, index: int, val: int) -> None:
        # If index is greater than the length, the node will not be inserted.
        if index > self.size:
            return
        # If index is negative, the node will be inserted at the head.
        if index < 0:
            index = 0
            
        # Find predecessor and successor of the node to be inserted
        if index < self.size - index:
            pred = self.head
            for _ in range(index):
                pred = pred.next
            succ = pred.next
        else:
            succ = self.tail
            for _ in range(self.size - index):
                succ = succ.prev
            pred = succ.prev
            
        self.size += 1
        to_add = ListNode(val)
        to_add.prev = pred
        to_add.next = succ
        pred.next = to_add
        succ.prev = to_add

    def deleteAtIndex(self, index: int) -> None:
        # Check for out-of-bounds indices
        if index < 0 or index >= self.size:
            return
        
        # Find predecessor and successor of the node to be deleted
        if index < self.size - index:
            pred = self.head
            for _ in range(index):
                pred = pred.next
            succ = pred.next.next
        else:
            succ = self.tail
            for _ in range(self.size - index - 1):
                succ = succ.prev
            pred = succ.prev.prev
            
        self.size -= 1
        pred.next = succ
        succ.prev = pred